presto-main/src/main/java/com/facebook/presto/operator/BigintInMemoryJoinHash.java

```diff
@@ -0,0 +1,232 @@
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int channelCount;
+    private final int mask;
+    private final int[] key;
+    private Optional<int[]> positionLinks = Optional.empty(); // lazy initialized
+    private final long totalMemoryUsage;
+    private final long[] values;
+    private Optional<BitSet> nullsMask = Optional.empty(); // lazy initialized
+
+    public BigintInMemoryJoinHash(LongArrayList addresses, PagesHashStrategy pagesHashStrategy, List<List<Block>> channels, List<Integer> joinChannels)
+    {
+        checkState(requireNonNull(joinChannels, "joinChannels is null").size() == 1, "Only single channel joins are supported");
+        int joinChannel = joinChannels.get(0);
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.pagesHashStrategy = requireNonNull(pagesHashStrategy, "pagesHashStrategy is null");
+        this.channelCount = pagesHashStrategy.getChannelCount();
+
+        // reserve memory for the arrays
+        int hashSize = HashCommon.arraySize(addresses.size(), 0.75f);
+
+        // This implementation assumes arrays used in the hash are always a power of 2
+        mask = hashSize - 1;
+        key = new int[hashSize];
+        Arrays.fill(key, -1);
+
+        values = new long[addresses.size()];
+
+        // First extract all values from blocks to native array.
+        // Somehow having this as a separate loop is much faster compared
+        // to extracting hashes on the fly in the loop below..
+        extractValues(addresses, channels.get(joinChannel));
+
+        for (int position = 0; position < addresses.size(); position++) {
+            long value = values[position];
+            int hashPos = getHashPosition(value, mask);
+
+            // look for an empty slot or a slot containing this key
+            int currentKey = key[hashPos];
+            while (currentKey != -1) {
+                long currentValue = values[currentKey];
+                if (currentValue == value && bothPositionsNullOrNonNull(currentKey, position)) {
+                    // found a slot for this key
+                    // link the new key position to the current key position
+                    setPositionLinks(position, currentKey);
+
+                    // key[pos] updated outside of this loop
+                    break;
+                }
+                // increment position and mask to handler wrap around
+                hashPos = (hashPos + 1) & mask;
+                //currentKey = key.get(hashPos);
+                currentKey = key[hashPos];
+            }
+
+            key[hashPos] = position;
+        }
+
+        totalMemoryUsage = sizeOf(key) + sizeOf(values)
+                +  sizeOf(addresses.elements()) + pagesHashStrategy.getSizeInBytes()
+                + (positionLinks.isPresent() ? sizeOf(positionLinks.get()) : 0)
+                + (nullsMask.isPresent() ? sizeOf(nullsMask.get().toLongArray()) : 0);
```

Comment:
from javadoc for `toLongArray`:
```
Returns a new long array containing all the bits in this bit set.
```

Maybe use `size` instead? BitSet uses long array internally (words) and store  6 bits per each word. Size returns `words.length * 6`. so here you would need to divide size by 6 and use sizeOfLongArray.